docs/proposals/component-registration.md

```diff
@@ -0,0 +1,406 @@
+    - [Self-Healing](#self-healing)
+  - [Design](#design)
+    - [Phased Changes](#phased-changes)
+    - [API Changes](#api-changes)
+      - [API Types](#api-types)
+    - [Kubectl (CLI)](#kubectl-cli)
+    - [Storage](#storage)
+    - [Status Updates](#status-updates)
+  - [Example](#example)
+  - [Alternate Naming](#alternate-naming)
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Terminology
+
+A Kubernetes Cluster **Component**...
+
+- Is considered a part of the cluster
+- Must register (or be registered) with Kubernetes (apiserver) to join the cluster
+- Must know about Kubernetes (speak the API)
+- Must be deployed/managed by the cluster admin/operator (rather than a user)
+- Should expose a method for remotely probing the liveness and readiness of each instance (e.g. tcp/http/https).
+- May be scheduled/deployed on the cluster (primary components vs addon components)
+- May be required for the cluster to function (required vs optional)
+- May be added/removed from a running cluster (mutable cluster component membership over time)
+- May require admin API access (privileged behavior)
+- May consist of one or more instances
+- May be deployed on different (network accessible) machines
+- May be upgraded over time
+
+The **External Components** are components that are deployed and lifecycle managed by some external system.
+
+The **Addon Components** are components that are deployed and lifecycle managed by the Kubernetes to which they provide functionality. These components must be optional in order for them to be deployable by Kubernetes (at least until an HA bootstrapping method is developed). Kubernetes conformance tests currently require some of these (e.g. kube-dns), but they are not required for orchestration and scheduling functionality.
+
+The **Primary Components** (currently apiserver, controller-manager, and scheduler) are components that are required for orchestration and scheduling functionality. These components are currently all External Components but may become Addon Components, once an HA bootstrapping method is developed.
+
+The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
+
+**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
+
+**Etcd** is a consistent key value storage cluster. It should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself with Kubernetes but must be configured. In general, the apiserver attempts to hide/abstract Etcd from the rest of the components with its API.
```

Comment:
I expect that k8s will frequently be used to host etcd clusters (even tho i usually recommend against ti, since they're more pets than cattle), but those are more frequently going to be for other applications, not k8s itself.

If at some point we figure out how to bootstrap from an external etcd to an internal etcd, we can deal with that change in the plan when it becomes feasible. Until then, I'm operating on the assumption that it is hard, and that most operators will want an external etcd.